dune-release publish sometimes fails

```
[?] Upload _build/mirage-entropy-v0.5.1.tbz as release asset? [Y/n]
[-] Uploading _build/mirage-entropy-v0.5.1.tbz as a release asset for v0.5.1 via github's API
dune-release: [ERROR] Could not extract archive url from:
                       {"message":"Error saving
                      asset","request_id":"D9BE:35EF:3D95:4F50:5E5401F7"}
```

And now the release is created on github, but the tarball is not uploaded. A `dune-release publish distrib` fails since it attempts to re-create the release.

To workaround, I delete the release and start `dune-release publish` again. Could there be a better way to deal with these errors? Could there be a more robust way to upload the tarball?

I suggest a `dune-release publish` subcommand which only uploads the tarball would be a good first step.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,7 @@
 
 - Add a `--token` option to `dune-release publish` and `dune-release opam` commands to specify a github token. This allows dune-release to be called through a Github Actions workflow and use the github token through an environment variable. (#284, @gpetiot)
 - Log curl calls on verbose/debug mode (#281, @gpetiot)
+- Try to publish the release asset again after it failed (#272, @gpetiot)
 
 ### Changed
 

--- a/lib/github.ml
+++ b/lib/github.ml
@@ -197,14 +197,18 @@ let curl_create_release ~token ~dry_run version msg user repo =
   run_with_auth ~dry_run ~default_body ~auth curl_t
   >>= Github_v3_api.Release_response.release_id
 
-let curl_upload_archive ~token ~dry_run archive user repo release_id =
+let curl_upload_archive ~token ~dry_run ~yes archive user repo release_id =
   let curl_t = Curl.upload_archive ~archive ~user ~repo ~release_id in
   github_auth ~dry_run ~user token >>= fun auth ->
   let default_body =
     `Assoc [ ("browser_download_url", `String D.download_url) ]
   in
-  run_with_auth ~dry_run ~default_body ~auth curl_t
-  >>= Github_v3_api.Upload_response.browser_download_url
+  Prompt.try_again ~yes ~default_answer:Prompt.Yes
+    ~question:(fun l ->
+      l "Uploading %a as release asset failed. Try again?" Text.Pp.path archive)
+    (fun () ->
+      run_with_auth ~dry_run ~default_body ~auth curl_t
+      >>= Github_v3_api.Upload_response.browser_download_url)
 
 let open_pr ~token ~dry_run ~title ~distrib_user ~user ~branch ~opam_repo body =
   let curl_t = Curl.open_pr ~title ~user ~branch ~body ~opam_repo in
@@ -280,7 +284,7 @@ let publish_distrib ?token ?distrib_uri ~dry_run ~msg ~archive ~yes p =
   App_log.status (fun l ->
       l "Uploading %a as a release asset for %a via github's API" Text.Pp.path
         archive Text.Pp.version tag);
-  curl_upload_archive ~token ~dry_run archive user repo id
+  curl_upload_archive ~token ~dry_run ~yes archive user repo id
 
 (*---------------------------------------------------------------------------
    Copyright (c) 2016 Daniel C. Bünzli

--- a/lib/prompt.ml
+++ b/lib/prompt.ml
@@ -33,6 +33,7 @@ let confirm_or_abort ~question ~yes ~default_answer =
 let rec try_again ~question ~yes ~default_answer f =
   match f () with
   | Ok x -> Ok x
-  | Error _ ->
+  | Error (`Msg err) ->
+      App_log.unhappy (fun l -> l "%s" err);
       confirm_or_abort ~yes ~question ~default_answer >>= fun () ->
       try_again ~question ~yes ~default_answer f

--- a/lib/prompt.ml
+++ b/lib/prompt.ml
@@ -1,5 +1,7 @@
 type answer = Yes | No
 
+open Bos_setup.R.Infix
+
 let ask f ~default_answer =
   let options : ('a, Format.formatter, unit, unit) format4 =
     match default_answer with Yes -> " [Y/n]" | No -> " [y/N]"
@@ -27,3 +29,10 @@ let confirm ~question ~yes ~default_answer =
 let confirm_or_abort ~question ~yes ~default_answer =
   if confirm ~question ~yes ~default_answer then Ok ()
   else Error (`Msg "Aborting on user demand")
+
+let rec try_again ~question ~yes ~default_answer f =
+  match f () with
+  | Ok x -> Ok x
+  | Error _ ->
+      confirm_or_abort ~yes ~question ~default_answer >>= fun () ->
+      try_again ~question ~yes ~default_answer f

--- a/lib/prompt.ml
+++ b/lib/prompt.ml
@@ -30,10 +30,11 @@ let confirm_or_abort ~question ~yes ~default_answer =
   if confirm ~question ~yes ~default_answer then Ok ()
   else Error (`Msg "Aborting on user demand")
 
-let rec try_again ~question ~yes ~default_answer f =
+let rec try_again ?(limit = 1) ~question ~yes ~default_answer f =
   match f () with
   | Ok x -> Ok x
-  | Error (`Msg err) ->
+  | Error (`Msg err) when limit > 0 ->
       App_log.unhappy (fun l -> l "%s" err);
       confirm_or_abort ~yes ~question ~default_answer >>= fun () ->
-      try_again ~question ~yes ~default_answer f
+      try_again ~limit:(limit - 1) ~question ~yes ~default_answer f
+  | Error x -> Error x

--- a/lib/prompt.mli
+++ b/lib/prompt.mli
@@ -19,11 +19,13 @@ val confirm_or_abort :
     [Error (`Msg "Aborting on user demand")] for no *)
 
 val try_again :
+  ?limit:int ->
   question:('a, unit) Logs.msgf ->
   yes:bool ->
   default_answer:answer ->
   (unit -> ('b, Rresult.R.msg) result) ->
   ('b, Rresult.R.msg) result
-(** [try_again ~question ~yes ~default_answer f] prompts the user to try running
-    [f] again if it failed, until [f] eventually succeeds or the user aborts the
-    process by answering no. *)
+(** [try_again ?limit ~question ~yes ~default_answer f] prompts the user to try
+    running [f] again if it failed, until [f] eventually succeeds or the user
+    aborts the process by answering no or the maximum number of retries [limit]
+    is reached (retries only once by default). *)

--- a/lib/prompt.mli
+++ b/lib/prompt.mli
@@ -17,3 +17,13 @@ val confirm_or_abort :
   (unit, Rresult.R.msg) result
 (** Same as [confirm] but returns [Ok ()] for yes and
     [Error (`Msg "Aborting on user demand")] for no *)
+
+val try_again :
+  question:('a, unit) Logs.msgf ->
+  yes:bool ->
+  default_answer:answer ->
+  (unit -> ('b, 'c) result) ->
+  ('b, Rresult.R.msg) result
+(** [try_again ~question ~yes ~default_answer f] prompts the user to try running
+    [f] again if it failed, until [f] eventually succeeds or the user aborts the
+    process by answering no. *)

--- a/lib/prompt.mli
+++ b/lib/prompt.mli
@@ -22,7 +22,7 @@ val try_again :
   question:('a, unit) Logs.msgf ->
   yes:bool ->
   default_answer:answer ->
-  (unit -> ('b, 'c) result) ->
+  (unit -> ('b, Rresult.R.msg) result) ->
   ('b, Rresult.R.msg) result
 (** [try_again ~question ~yes ~default_answer f] prompts the user to try running
     [f] again if it failed, until [f] eventually succeeds or the user aborts the

